src/fluids/rapier_fluid.rs

```diff
@@ -133,34 +118,43 @@ impl RapierFluid {
     }
 
     pub fn delete_points(&mut self, indices: Vec<i32>, physics_engine: &mut PhysicsEngine) {
-        let removals = indices.clone().into_iter().collect::<VecDeque<_>>();
-        for index in removals {
-            self.points.remove(index as usize);
-            self.velocities.remove(index as usize);
-            self.accelerations.remove(index as usize);
-        }
+        // DO NOT modify local cache (self.points, self.velocities) here.
+        // Doing so causes a race condition, as the physics_engine call
+        // only *schedules* the deletion for the next physics step.
+        // The local cache will be correctly updated on the next get_points()
+        // or get_velocities() call.
         physics_engine.fluid_delete_points(self.space_id, self.fluid_handle, indices);
     }
 
-    pub fn get_points(&mut self, physics_engine: &mut PhysicsEngine) -> &Vec<Vector> {
+    // The individual get_... methods were desynchronizing the local cache.
+    // get_points() would update self.points, but not self.velocities, etc.
+    // This new function syncs all particle data at once.
+    fn sync_cache_from_engine(&mut self, physics_engine: &mut PhysicsEngine) {
         if self.is_valid() {
             self.points = physics_engine.fluid_get_points(self.space_id, self.fluid_handle);
+            self.velocities = physics_engine.fluid_get_velocities(self.space_id, self.fluid_handle);
+            self.accelerations =
+                physics_engine.fluid_get_accelerations(self.space_id, self.fluid_handle);
         }
+    }
+
+    pub fn get_points(&mut self, physics_engine: &mut PhysicsEngine) -> &Vec<Vector> {
+        // Now, requesting points syncs everything, guaranteeing the
+        // local vectors (points, velocities, accelerations) have the same length.
+        self.sync_cache_from_engine(physics_engine);
```

Comment:
Is it enough to call it here? Should we instead call it after the step function? Reason being that in case the fluid gets deleted and recreated, these would persist.